internal/expr/parser.go

```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
 	return "", false
 }
 
+// parseColumn parses a column made up of name bytes, optionally dot-prefixed by
+// its table name. parseColumn returns an error so that it can be used with
+// parseList.
+func (p *Parser) parseColumn() (fullName, bool, error) {
+	cp := p.save()
+
+	if id, ok := p.parseIdentifierAsterisk(); ok {
+		if id != "*" && p.skipByte('.') {
+			if idCol, ok := p.parseIdentifierAsterisk(); ok {
+				return fullName{prefix: id, name: idCol}, true, nil
+			}
+		} else {
+			// A column name specified without a table prefix should be in name.
+			return fullName{name: id}, true, nil
+		}
+	}
+
+	cp.restore()
+	return fullName{}, false, nil
+}
+
+func (p *Parser) parseTarget() (fullName, bool, error) {
+	cp := p.save()
+
+	if p.skipByte('&') {
+		return p.parseGoFullName()
+	}
+
+	cp.restore()
+	return fullName{}, false, nil
+}
+
 // parseGoFullName parses a Go type name qualified by a tag name (or asterisk)
-// of the form "TypeName.col_name". On success it returns the parsed fullName,
+// of the form "&TypeName.col_name". On success it returns the parsed FullName,
 // true and nil. If a Go full name is found, but not formatted correctly, false
 // and an error are returned. Otherwise the error is nil.
 func (p *Parser) parseGoFullName() (fullName, bool, error) {
 	cp := p.save()
+
 	if id, ok := p.parseIdentifier(); ok {
-		if p.skipByte('.') {
-			if idField, ok := p.parseIdentifierAsterisk(); ok {
-				return fullName{id, idField}, true, nil
-			}
-			return fullName{}, false,
-				fmt.Errorf("invalid identifier near char %d", p.pos)
+		if !p.skipByte('.') {
+			return fullName{}, false, fmt.Errorf("go object near char %d not qualified", p.pos)
+		}
+
+		idField, ok := p.parseIdentifierAsterisk()
+		if !ok {
+			return fullName{}, false, fmt.Errorf("invalid identifier near char %d", p.pos)
 		}
-		return fullName{}, false,
-			fmt.Errorf("go object near char %d not qualified", p.pos)
+		return fullName{id, idField}, true, nil
 	}
+
 	cp.restore()
 	return fullName{}, false, nil
 }
 
-// parseInputExpression parses an input expression of the form $Type.name.
+// parseList takes a parsing function that returns a fullName and parses a
+// bracketed, comma seperated, list.
+func (p *Parser) parseList(parseFn func(p *Parser) (fullName, bool, error)) ([]fullName, bool, error) {
+	cp := p.save()
```

Comment:
The save can be moved past the first skip below, as a small optimization that prevents doing anything unless this is indeed looking like a proper start at all.

Reply:
I've just realised we can't do this. Because the skipByte is negated, if we **don't** exit the function at the if then the state will have been changed, we will have skipped the `(` byte. So we would be saving the state after the byte has been skipped, thus when we restore it it would not be a true restore.

Reply:
Good catch. :+1: